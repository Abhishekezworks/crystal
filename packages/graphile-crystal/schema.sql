create extension if not exists citext;
create extension if not exists pgcrypto;

drop schema if exists app_public cascade;

create schema app_public;

create table app_public.users (
  id uuid primary key default gen_random_uuid(),
  username citext not null unique,
  gravatar_url text,
  created_at timestamptz not null default now()
);

create table app_public.forums (
  id uuid primary key default gen_random_uuid(),
  name text not null
);

create table app_public.messages (
  id uuid primary key default gen_random_uuid(),
  forum_id uuid not null references app_public.forums,
  author_id uuid not null references app_public.users,
  body text not null,
  created_at timestamptz not null default now()
);

insert into app_public.users (username) values
  ('Alice'),
  ('Bob'),
  ('Carrie');

insert into app_public.forums (name) values
  ('Cats'),
  ('Dogs'),
  ('Postgres');

insert into app_public.messages (forum_id, author_id, body)
  select forums.id, users.id, forums.name || ' = awesome -- ' || users.username
  from app_public.users, app_public.forums;
