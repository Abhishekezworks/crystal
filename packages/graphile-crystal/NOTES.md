# What to do?

`yarn ts-node examples/example2.ts`

# Current status

Currently trying to make sense of connections and cursors. Current thinking is
that the connection gets a "template" plan, then the
edges/nodes/aggregates/pageInfo/etc can clone this plan and augment it as
necessary. How to construct the cursor though? I'm making this a
PgClassSelectPlan concern via `cursor()` method but this isn't really right
because scalar connections can also have cursors. Cursors have to factor in the
orderBy which could be arbitrary data, even from joins, so has to be complex SQL
probably.

Also plans will probably need to never override the constructor.
