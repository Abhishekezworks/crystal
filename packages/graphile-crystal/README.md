# graphile-crystal

GraphQL's planning problem: solved.

## Restrictions

For Graphile Crystal to work, you need to adhere to the following restrictions:

- schema:
  - every resolver must be wrapped in crystal; this can be achieved manually, or
    by calling `enforceCrystal(schema)`
- during execution:
  - `context` must be an object (anything suitable to be used as the key to a
    `WeakMap`)
  - `rootValue` must be an object or `null`/`undefined`
  - `variables` must be an object or `null`/`undefined` (which GraphQL requires
    anyway)
  - None of the above may be a scalar
  - Do not provide `fieldResolver` (i.e. don't try and override the default
    field resolver)

## Advice

To reap the most benefit from using Graphile Crystal, you want as little to
change between executions as possible. In particular, this means you should:

- Cache (e.g. with a LRU cache) the parsed GraphQL document, so the same AST can
  be reused over and over for the same document
- Cache (e.g. with a LRU cache) the GraphQL `context` object, so the same
  context can be reused over and over for the same user
- Don't use `rootValue` (Do you really need it? Use `context` instead.)
- Where possible, memoize the variables object (e.g. using a cache over
  `canonicalJSONStringify(variables)`) so the same variables results in the same
  object in memory

## API

### `enforceCrystal(schema)`

Ensures every resolver in `schema` is wrapped in crystal. BEWARE - do not do
this if resolvers may already be wrapped, dual-wrapping resolvers will result in
hard-to-find bugs. This is particularly relevant if you're using other libraries
that modify a schema after it was created, such as `graphql-shield`.

TODO: mutate or return derivative?
