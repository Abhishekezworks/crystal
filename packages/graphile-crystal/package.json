{
  "name": "graphile-crystal",
  "version": "4.7.0",
  "description": "Look-ahead engine for GraphQL",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "jest -i",
    "prepack": "tsc -b",
    "md": "spec-md CRYSTAL_FLOW.md > CRYSTAL_FLOW.html"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphile/graphile-build.git"
  },
  "keywords": [
    "graphile",
    "graphql",
    "engine",
    "apollo",
    "graphqljs",
    "plugin",
    "build",
    "extension"
  ],
  "author": "Benjie Gillam <code@benjiegillam.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/graphile/graphile-build/issues"
  },
  "homepage": "https://graphile.org/graphile-build/",
  "dependencies": {
    "@graphile/lru": "4.6.0-alpha.0",
    "@types/node": "^14.14.44",
    "iterall": "^1.3.0",
    "lodash": "^4.17.15",
    "strip-ansi": "6.x",
    "tslib": "^2.2.0"
  },
  "engines": {
    "node": ">=14.17"
  },
  "peerDependencies": {
    "graphql": "*"
  },
  "files": [
    "dist",
    "index.js"
  ],
  "devDependencies": {
    "@types/jest": "^25.2.3",
    "@types/lodash": "^4.14.151",
    "@types/nodemon": "1.19.0",
    "graphql": "npm:graphql@experimental-stream-defer",
    "jest": "^26.0.1",
    "nodemon": "^2.0.7",
    "pg-sql2": "4.7.0",
    "spec-md": "^3.0.2",
    "ts-node": "^9.1.1",
    "typescript": "^4.2.4"
  }
}
