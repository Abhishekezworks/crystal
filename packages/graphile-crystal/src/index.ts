import debugFactory from "debug";

import { crystalPrint, crystalPrintPathIdentity } from "./crystalPrint";

debugFactory.formatters.c = crystalPrint;
debugFactory.formatters.p = crystalPrintPathIdentity;

export { Aether } from "./aether";
export { ROOT_PATH } from "./constants";
export { defer, Deferred } from "./deferred";
export { crystalEnforce } from "./enforceCrystal";
export {
  InputListPlan,
  InputObjectPlan,
  InputPlan,
  InputStaticLeafPlan,
} from "./input";
export {
  ArgumentPlanResolver,
  BaseGraphQLArguments,
  BaseGraphQLContext,
  BaseGraphQLRootValue,
  BaseGraphQLVariables,
  CrystalResultsList,
  CrystalResultStreamList,
  CrystalValuesList,
  ExecutablePlanResolver,
  GraphileCrystalFieldConfig,
  InputObjectFieldPlanResolver,
  PolymorphicData,
} from "./interfaces";
export {
  BasePlan,
  ExecutablePlan,
  isExecutablePlan,
  isModifierPlan,
  isObjectLikePlan,
  ModifierPlan,
  ObjectLikePlan,
  PolymorphicPlan,
} from "./plan";
export {
  __ListItemPlan,
  __TrackedObjectPlan,
  __ValuePlan,
  access,
  AccessPlan,
  aether,
  assertListCapablePlan,
  constant,
  ConstantPlan,
  context,
  debugPlans,
  each,
  EachPlan,
  first,
  FirstPlan,
  lambda,
  LambdaPlan,
  list,
  ListCapablePlan,
  ListPlan,
  makeMapper,
  map,
  MapPlan,
  object,
  ObjectPlan,
  reverse,
  reverseArray,
  ReversePlan,
} from "./plans";
export { polymorphicWrap, resolveType } from "./polymorphic";
export {
  $$crystalWrapped,
  crystalWrapResolve,
  crystalWrapSubscribe,
} from "./resolvers";
export {
  arraysMatch,
  inputObjectFieldSpec,
  inputObjectSpec,
  objectFieldSpec,
  objectSpec,
  planGroupsOverlap,
} from "./utils";

// Handy for debugging
export { getCurrentParentPathIdentity } from "./global";
