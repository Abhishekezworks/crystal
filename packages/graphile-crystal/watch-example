GRAPHILE_ENV=development DEBUG="crystal:*" yarn nodemon --watch dist dist/examples/example.js
