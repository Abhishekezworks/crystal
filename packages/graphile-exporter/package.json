{
  "name": "graphile-exporter",
  "version": "0.0.1",
  "description": "Take a GraphQL schema from memory and write it to file",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js"
    },
    "./helpers": {
      "types": "./dist/helpers.d.ts",
      "import": "./dist/helpers.js"
    }
  },
  "scripts": {
    "test": "jest -i",
    "prepack": "tsc -b"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphile/graphile-build.git"
  },
  "keywords": [
    "graphql",
    "schema",
    "memory",
    "export",
    "output",
    "write",
    "file",
    "graphile"
  ],
  "author": "Benjie Gillam <code@benjiegillam.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/graphile/graphile-build/issues"
  },
  "homepage": "https://graphile.org/",
  "dependencies": {
    "@babel/generator": "^7.16.0",
    "@babel/parser": "^7.16.3",
    "@babel/template": "^7.16.0",
    "@types/node": "^16.11.6",
    "graphile-crystal": "4.7.0",
    "tslib": "^2.3.1"
  },
  "engines": {
    "node": ">=16"
  },
  "peerDependencies": {
    "graphql": "*"
  },
  "files": ["dist"],
  "devDependencies": {
    "@types/babel__generator": "^7",
    "@types/babel__template": "^7",
    "@types/jest": "^27.0.2",
    "graphql": "npm:graphql@experimental-stream-defer",
    "jest": "^27.3.1",
    "ts-node": "^10.4.0",
    "typescript": "next"
  }
}
