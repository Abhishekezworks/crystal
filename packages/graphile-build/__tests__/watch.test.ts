import { EventEmitter } from "events";
import { lambda, object } from "graphile-crystal";
import { EXPORTABLE } from "graphile-exporter";
import type { ExecutionResult, GraphQLSchema } from "graphql";
import {
  graphql,
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  printSchema,
} from "graphql";

import {
  defaultPlugins as allDefaultPlugins,
  getBuilder,
  MutationPlugin,
} from "../src";

function isAsyncIterable(
  t: ExecutionResult<any> | AsyncIterable<any>,
): t is AsyncIterable<any> {
  return Symbol.asyncIterator in t;
}

const options = {};

const defaultPlugins = allDefaultPlugins.filter(
  (plugin) => plugin !== MutationPlugin,
);

declare global {
  namespace GraphileEngine {
    interface Build {
      dummyCounter: number;
    }
    interface ScopeGraphQLObjectTypeFieldsField {
      isDummyField?: boolean;
    }
  }
}

const makePluginEtc = (defaultCounter = 0) => {
  let counter = defaultCounter;

  const eventEmitter = new EventEmitter();

  const DummyWatchPlugin: GraphileEngine.Plugin = async (builder) => {
    builder.registerWatcher(
      (triggerRebuild) => {
        eventEmitter.on("change", triggerRebuild);
      },
      (triggerRebuild) => {
        eventEmitter.removeListener("change", triggerRebuild);
      },
    );

    builder.hook("build", (build) => {
      return build.extend(
        build,
        {
          dummyCounter: counter,
        },
        "TEST",
      );
    });

    builder.hook("init", (_, build) => {
      const { dummyCounter } = build;
      build.registerObjectType(
        `Dummy${dummyCounter}`,
        {},
        null,
        () => ({
          fields: () => {
            return {
              n: {
                type: new GraphQLNonNull(GraphQLInt),
                plan: EXPORTABLE(
                  (dummyCounter, lambda) => () =>
                    lambda(null, () => dummyCounter),
                  [dummyCounter, lambda],
                ),
              },
            };
          },
        }),
        "watch.test.ts",
      );
      return _;
    });

    builder.hook("GraphQLObjectType:fields", (fields, build, context) => {
      const { dummyCounter, extend, getTypeByName } = build;
      const {
        scope: { isRootQuery },
        fieldWithHooks,
      } = context;
      if (!isRootQuery) return fields;
      const Dummy = getTypeByName(`Dummy${dummyCounter}`);
      return extend(
        fields,
        {
          dummy: fieldWithHooks(
            { fieldName: "dummy", isDummyField: true },
            () => {
              return {
                type: Dummy,
                plan: EXPORTABLE(
                  (object) =>
                    function plan() {
                      return object({});
                    },
                  [object],
                ),
              };
            },
          ),
        },
        "TEST",
      );
    });
  };

  return {
    plugin: DummyWatchPlugin,
    eventEmitter,
    setN(n: number) {
      counter = n;
    },
  };
};

test("generated schema n = 0, n = 3", async () => {
  const schema0 = (
    await getBuilder([...defaultPlugins, makePluginEtc(0).plugin])
  ).buildSchema();
  expect(schema0).toMatchSnapshot();
  const schema3 = (
    await getBuilder([...defaultPlugins, makePluginEtc(3).plugin])
  ).buildSchema();
  expect(schema3).toMatchSnapshot();
});

test("schema is cached if no watcher fires", async () => {
  const { plugin, setN } = makePluginEtc();
  const builder = await getBuilder([...defaultPlugins, plugin], options);
  await builder.watchSchema();

  const schema0 = builder.buildSchema();
  const schema0_2 = builder.buildSchema();
  setN(70);
  const schema0_3 = builder.buildSchema();
  expect(schema0).toBe(schema0_2);
  expect(schema0).toBe(schema0_3);

  await builder.unwatchSchema();
});

test("schema is equivalent (but not identical) if rebuild fires but no changes occur", async () => {
  const { plugin, eventEmitter } = makePluginEtc();
  const builder = await getBuilder([...defaultPlugins, plugin], options);
  await builder.watchSchema();

  const schema0 = builder.buildSchema();
  eventEmitter.emit("change");
  const schema0_2 = builder.buildSchema();

  expect(schema0).not.toBe(schema0_2);
  expect(printSchema(schema0)).toEqual(printSchema(schema0_2));

  await builder.unwatchSchema();
});

test("schema is updated when rebuild triggered", async () => {
  const { plugin, setN, eventEmitter } = makePluginEtc();
  const builder = await getBuilder([...defaultPlugins, plugin], options);
  await builder.watchSchema();

  const schema0 = builder.buildSchema();
  setN(70);
  eventEmitter.emit("change");
  const schema1 = builder.buildSchema();

  expect(schema0).not.toEqual(schema1);
  expect(printSchema(schema0)).not.toEqual(printSchema(schema1));

  const getNFrom = async (schema: GraphQLSchema) => {
    const result = await graphql({
      schema,
      source: `
        query {
          dummy {
            n
          }
        }
      `,
    });
    if (isAsyncIterable(result)) {
      throw new Error("Did not expect an async iterator here");
    }
    if ("errors" in result && result.errors) {
      // eslint-disable-next-line no-console
      console.log(result.errors.map((e) => e.originalError));
    }
    expect(result.errors).toBeFalsy();
    return (result.data as any)?.dummy.n;
  };

  expect(await getNFrom(schema0)).toEqual(0);
  expect(await getNFrom(schema1)).toEqual(70);

  await builder.unwatchSchema();
});
