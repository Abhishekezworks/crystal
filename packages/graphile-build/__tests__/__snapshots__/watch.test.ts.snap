// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated schema n = 0, n = 3 1`] = `
type Dummy0 {
  n: Int!
}

"""The root query type which gives access points into the data universe."""
type Query {
  dummy: Dummy0
}
`;

exports[`generated schema n = 0, n = 3 2`] = `
type Dummy3 {
  n: Int!
}

"""The root query type which gives access points into the data universe."""
type Query {
  dummy: Dummy3
}
`;
